include/xtensor/xindexview.hpp

```diff
@@ -0,0 +1,726 @@
+    inline auto xindexview<T, S>::storage_end() const -> const_storage_iterator
+    {
+        return cend();
+    }
+    //@}
+
+    /**
+     * Returns a constant iterator to the first element of the buffer
+     * containing the elements of the view.
+     */
+    template <class T, class S>
+    inline auto xindexview<T, S>::storage_cbegin() const -> const_storage_iterator
+    {
+        return cbegin();
+    }
+
+    /**
+     * Returns a constant iterator to the element following the last
+     * element of the buffer containing the elements of the view.
+     */
+    template <class T, class S>
+    inline auto xindexview<T, S>::storage_cend() const -> const_storage_iterator
+    {
+        return cend();
+    }
+    //@}
+
+    /*************************************
+     * xindexview_stepper implementation *
+     *************************************/
+
+    template <bool is_const, class T, class S>
+    inline xindexview_stepper<is_const, T, S>::xindexview_stepper(view_type* view, const shape_type& shape) noexcept
+        : p_view(view), m_shape(shape), m_index(make_sequence<index_type>(shape.size(), size_type(0)))
+    {
+    }
+
+    template <bool is_const, class T, class S>
+    inline void xindexview_stepper<is_const, T, S>::step(size_type dim, size_type n)
+    {
+        m_index[dim] += n;
+    }
+
+    template <bool is_const, class T, class S>
+    inline void xindexview_stepper<is_const, T, S>::step_back(size_type dim, size_type n)
+    {
+        m_index[dim] -= 1;
+    }
+
+    template <bool is_const, class T, class S>
+    inline void xindexview_stepper<is_const, T, S>::reset(size_type dim)
+    {
+        m_index[dim] = 0;
+    }
+
+    template <bool is_const, class T, class S>
+    inline void xindexview_stepper<is_const, T, S>::to_end()
+    {
+        m_index = m_shape;
+    }
+
+    template <bool is_const, class T, class S>
+    inline bool xindexview_stepper<is_const, T, S>::equal(const self_type& rhs) const
+    {
+        return p_view == rhs.p_view && std::equal(m_index.begin(), m_index.end(), rhs.m_index.begin());
+    }
+
+    template <bool is_const, class T, class S>
+    inline auto xindexview_stepper<is_const, T, S>::operator*() const -> reference
+    {
+        return (*p_view)[m_index];
+    }
+
+    template <bool is_const, class T, class S>
+    inline bool operator==(const xindexview_stepper<is_const, T, S>& it1,
+                           const xindexview_stepper<is_const, T, S>& it2)
+    {
+        return it1.equal(it2);
+    }
+
+    template <bool is_const, class T, class S>
+    inline bool operator!=(const xindexview_stepper<is_const, T, S>& it1,
+                           const xindexview_stepper<is_const, T, S>& it2)
+    {
+        return !(it1.equal(it2));
+    }
+
+    template <class T>
+    auto inline make_xindexview(T& arr, const std::vector<xindex>&& indices)
+    {
+        return xt::xindexview<T, std::vector<std::size_t>>(arr, std::move(indices));
```

Comment:
Same remark as for the xindex_view constructor, you can't mix const and rvalue reference. You can't use universal references if you want to allow initializer lists, so you'll need overloads for rvalue reference and const lvalue reference.